FPGA: Add new sample `SVD` (#2320)
oneapi-src/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/README.md

> # `Singular Value Decomposition` Sample

Reviewer: ```suggestion
# `Singular Value Decomposition (SVD)` Sample
```
I guess the other code samples include abbreviations so we should be consistent

| print_matrix.hpp       | Host code                       | Helper functions to print matrices.                                                                                                                                                                  |
| golden_pca.hpp         | Host code                       | A CPU reference PCA design that is used to calculate reference eigenvalues for the test bench.                                                                                                      |
| svd.hpp                | Host code with some device code | Contains wrapper function `SingularValueDecomposition` that launches individual kernels of the SVD design.                                                                                           |
| memory_transfers.hpp   | Device code                     | Contains kernel implementations of `MatrixReadFromDDRTo2PipesByBlocks` , `MatrixReadPipeToDDR` and `VectorReadPipeToDDR`. These kernels transfers data between DDR and streaming interfaces (pipes).|

> | usv_from_eigens.hpp       | Device code                     | Contains kernel implementation of `USVFromEigens`.                                                                                                                                                   |

Reviewer: You should add references to the shared files in the `include` directory to this table, but make it clear that those files are shared.

### Input covariance matrix computation

> The covariance computation in this design is the same as used in the [PCA](../pca/README.md) reference design except without standardization.

Reviewer: mention that `fpga_linalg::StreamingCovarianceMatrix` is re-used to obtain this computation

The covariance of input A is equal to the transpose of A multiplied by A:
```math
Cov(A) = A^{T}A
```

> Therefore this kernel performs a matrix multiplication by blocks, as described in the PCA reference design.

Reviewer: ```suggestion
Therefore this kernel performs a matrix multiplication by blocks, as described in the 
[PCA](../pca/) reference design, which also uses this functor.
```

### Eigenvalues and Eigenvectors computation

> We are reusing the `fpga_linalg::StreamingEigen` design from the PCA sample to compute eigenvalues and eigenvectors of the input covariance matrix. These eigenvalues and eigenvectors are going to be used to construct the final outputs of SVD.

Reviewer: ```suggestion
This design imports `streaming_eigen.hpp` from the shared [include]() directory. The eigenvalues and eigenvectors of the input covariance matrix are computed using the `fpga_linalg::StreamingEigen` functor. These eigenvalues and eigenvectors are going to be used to construct the final outputs of SVD.
```

use consistent language across each of these paragraphs so that the user is not surprised. The flow should be something like, "This design includes XXX. It uses YYYY functor to implement ZZZ functionality." Optionally, say "This functor is also used in AAAA design."

This kernel works only with input of rank sufficient matrix (all columns are linearly independent). It produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result of the SVD is known to be incorrect.

The source code for the the eigenvalues and eigenvectors computation kernel can be found in `streaming_eigen.hpp` in the shared [include](../../include) directory. An detailed explanation of the algorithm of computing eigenvalues can be found [here](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.

> ### Construct Output from Eigenvalues and Eigenvectors

Reviewer: (I assume) that the computations in this section are all found in post_process.hpp; please state explicitly state this so the reader knows where to look.

> The source code for the the eigenvalues and eigenvectors computation kernel can be found in `streaming_eigen.hpp` in the shared [include](../../include) directory. An detailed explanation of the algorithm of computing eigenvalues can be found [here](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.

Reviewer: ```suggestion
The [PCA](../pca/) reference design also uses this functor. A detailed explanation of the algorithm of computing eigenvalues can be found in [Eigen values and Eigen vectors computation](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.
```

Eric tells me it's bad form to use 'click here' links (he has rebuked me for this in the past). It's better to either build it into the text, or refer to the title of the thing you're linking to.

> In `usv_from_eigens.hpp`, the final outputs of SVD is constructed from the eigenvalues and eigenvectors of the input covariance. 

Reviewer: ```suggestion
The functor `USVFromEigens` in `usv_from_eigens.hpp`, constructs the final outputs of SVD from the eigenvalues and eigenvectors of the input covariance matrix. 
```

```math
A = USV^T
```
Since the $V$ matrix is orthogonal by construction, $V^{-1}$ is the same as $V^T$, therefore:
```math
AV = US
```
And since $S$ is a diagonal matrix, $S^{-1}$ is the same as multiplying by the reciprocals of each element on the main diagonal of $S$.

In summary, the $m \times n$ portion of matrix $U$ ($U_{[0:m][0:n]}$) can be obtained through:
![](assets/U_matrix.png)

When the input matrix $A$ is not square, the number of eigenvectors calculated is less than the number of columns in the $U$ matrix. An orthogonalization kernel is needed to generate extra orthogonal vectors.

### $U$ Matrix Orthogonalization
As mentioned above, when extra filler vectors are needed to complete the $U$ matrix, we need to orthogonalize the matrix.

An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md), so here we will insert an instance of the `fpga_linalg::streamingQRD` design. The $Q$ output of this kernel is orthogonalized $U$ matrix. Since we only care about the orthogonalized $U$ matrix, $R$ output of the Streaming QRD kernel is discarded.

### Demonstration Testbench

> In this sample, a testbench is used to demonstrate the SVD design.

Reviewer: should mention functor and filename.

| Optimized for        | Description
|:---                  |:---
| OS                   | Ubuntu* 20.04 <br> RHEL*/CentOS* 8 <br> SUSE* 15 <br> Windows* 10 <br> Windows Server* 2019
| Hardware             | Intel® Agilex® 7, Arria® 10, and Stratix® 10 FPGAs
| Software             | Intel® oneAPI DPC++/C++ Compiler

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> For using the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

### Performance

Performance results are based on testing as of April 12, 2024 with fixed 55 iterations.

> Using BSP: Agilex_ref 23.3

Reviewer 2: ```suggestion
```

>  ```
## Run the `SVD` Design

### Configurable Parameters

| Argument  | Description
|:---       |:---
| `<num>`   | (Optional) Specifies the number of times to repeat the decomposition. Its default value is **1** for the emulation and simulation flow and **16384** for the FPGA flow.

### On Linux

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./svd.fpga_emu
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./svd.fpga_sim
   ```

#### Run on FPGA

1. Run the sample on the FPGA device (only if you ran `cmake` with `-DFPGA_DEVICE=<board-support-package>:<board-variant>`).
   ```
   ./svd.fpga
   ```

### On Windows

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   svd.fpga_emu.exe
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   svd.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```

## Example Output

Example Output when running on the **Terasic DE10-Agilex Development Board**.

```

> Running on device: de10_agilex : Agilex Reference Platform (aclde10_agilex0)

Reviewer 2: Run this sample on N6001 to update the readme from DE10 to N6001 as we moved to this board.

This design imports `streaming_covariance_matrix.hpp` from the shared [include](../../include) directory. The covariance computation in this design uses the `fpga_linalg::StreamingCovarianceMatrix` functor, with the `standardized` parameter of the `StreamingCovarianceMatrix` functor set to `false`.

The covariance of input A is equal to the transpose of A multiplied by A:
```math
Cov(A) = A^{T}A
```
Therefore this kernel performs a matrix multiplication by blocks, as described in the [PCA](../pca/) reference design, which also uses this functor.

### Eigenvalues and Eigenvectors computation
This design imports `streaming_eigen.hpp` from the shared [include]() directory. The eigenvalues and eigenvectors of the input covariance matrix are computed using the `fpga_linalg::StreamingEigen` functor. These eigenvalues and eigenvectors are going to be used to construct the final outputs of SVD.

This kernel works only with input of rank sufficient matrix (all columns are linearly independent). It produces an output `rank_deficent_flag` to indicate if the input matrix is not linearly independent. If the flag is set to 1, the result of the SVD is known to be incorrect.

> The [PCA](../pca/) reference design also uses this functor. A detailed explanation of the algorithm of computing eigenvalues can be found in [Eigen values and Eigen vectors computation](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.

Reviewer: ```suggestion
The [PCA](../pca/) reference design also uses this functor. A detailed explanation of the algorithm of computing eigenvalues can be found in [Eigenvalues and Eigenvectors computation](../pca/README.md#eigen-values-and-eigen-vectors-computation) in the PCA sample.
```
Yohann needs to fix this spelling mistake in his code samples, but I won't put that on you in this PR.

```math
AV = US
```
And since $S$ is a diagonal matrix, $S^{-1}$ is the same as multiplying by the reciprocals of each element on the main diagonal of $S$.

In summary, the $m \times n$ portion of matrix $U$ ($U_{[0:m][0:n]}$) can be obtained through:
![](assets/U_matrix.png)

When the input matrix $A$ is not square, the number of eigenvectors calculated is less than the number of columns in the $U$ matrix. An orthogonalization kernel is needed to generate extra orthogonal vectors.

### $U$ Matrix Orthogonalization
As mentioned above, when extra filler vectors are needed to complete the $U$ matrix, we need to orthogonalize the matrix.

An efficient algorithm to do this is already implemented in our [QR Decomposition sample](../qrd/README.md), so here we will insert an instance of the `fpga_linalg::streamingQRD` functor from [`streaming_qrd.hpp`](../../include/streaming_qrd.hpp). The $Q$ output of this kernel is orthogonalized $U$ matrix. Since we only care about the orthogonalized $U$ matrix, $R$ output of the Streaming QRD kernel is discarded.

### Demonstration Testbench
In this sample, a testbench is used to demonstrate the SVD design.

> The resulting singular values are checked against eigen values calculated by a reference PCA algorithm that is also used in the PCA sample.

Reviewer: ```suggestion
The resulting singular values are checked against eigenvalues calculated by a reference PCA algorithm that is also used in the PCA sample.
```

### Performance

Performance results are based on testing as of September 12, 2024 with fixed 55 [QR iterations](../pca/README.md#eigen-values-and-eigen-vectors-computation).

> **Note**: Refer to the [Performance Disclaimers](/DirectProgramming/C++SYCL_FPGA/README.md#performance-disclaimers) section for important performance information.

| Device                      | Input size (Rows x Cols) | Throughput     |
|-----------------------------|--------------------------|----------------|
| Silicom FPGA SmartNIC N6011 | 32 x 32                  | 987 matrices/s |

> | Silicom FPGA SmartNIC N6011 | 128 x 32                 | 560 matrices/s |

Reviewer 2: Can you also add the numbers you got using numpy, giving specs of the machine you ran the test on?

Reviewer: I don't think we should report on the numpy numbers, since Intel legal requires high standards when reporting benchmark results on other peoples' tools.
